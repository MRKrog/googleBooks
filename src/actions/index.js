export const setError = (error) => ({
  type: "SET_ERROR",
  error
})

export const setLoading = (loading) => ({
  type: "SET_LOADING",
  loading
})

export const setBookSearch = (books) => ({
  type: "BOOK_SEARCH",
  books
})
