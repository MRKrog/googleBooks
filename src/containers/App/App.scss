.App {
  background-color: #282c3466;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;

  .modal-button {
    opacity: 1;
    border: none;
    position: fixed;
    bottom: 1em;
    right: 1em;
    font-size: 1em;
    font-weight: 100;
    letter-spacing: .04em;
    font-family: $fontOswald;
    background: #2196F3;
    color: #fff;
    z-index: 99;
    line-height: 1;
    border-radius: 5px;
    padding: .5em 1em;
    box-shadow: 1px 1px 8px 2px #000000c7;
    cursor: pointer;
    transition: all .5s ease;

    &:hover {
      opacity: .95;
    }

    i {
      font-weight: 100;
      opacity: .8;
    }
  }
}
