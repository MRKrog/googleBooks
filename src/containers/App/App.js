import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Search } from '../Search/Search';
import BookDisplay from '../BookDisplay/BookDisplay';
import Modal from '../Modal/Modal';
import { Loading } from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';
import { cleanBooks } from '../../utility/cleanBooks'

import * as actions from '../../actions/index';

export class App extends Component {
  constructor(props) {
    super();
    this.state = {
      modal: false,
    }
  }

  handleSearch = async (search) => {
    const { setBookSearch, setLoading, savedBooks } = this.props;
    const url = "https://www.googleapis.com/books/v1/volumes";
    const encodedSearch = encodeURI(search);
    const updatedTerm = this.checkQuery(encodedSearch)
    setLoading(true)
    try {
      const response = await fetch(`${url}?q=${updatedTerm}`)
      if(!response.ok) { throw new Error('Fetch Call Cannot Be Made')}
      const data = await response.json()
      let updatedBooks = cleanBooks(data.items, savedBooks)
      await setBookSearch(updatedBooks)
    } catch(error){
      this.props.setError(error.message)
    }
    setLoading(false)
  }

  checkQuery = (search) => {
    if(search.includes('&')){
      search = search.replace(/&/g, 'and')
    }
    return search
  }

  showModal = () => {
    this.setState({
      modal: true
    })
  }

  hideModal = () => {
    this.setState({
      modal: false
    })
  }

  render() {
    const { loading, searchedBooks, savedBooks } = this.props;
    const { modal } = this.state;
    return (
      <div className="App">
      { savedBooks.length > 0 &&
        <button className="modal-button" onClick={this.showModal}>
          View Saved Books <i className="fas fa-eye"></i>
        </button>
      }
      <Search handleSearch={this.handleSearch} />
      {
        loading ?
        <Loading /> :
        <div className="BookResults">
          {
            searchedBooks.length > 0 &&
            <BookDisplay />
          }
        </div>
      }
      {
        modal &&
        <Modal hideModal={this.hideModal} />
      }
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  error: state.error,
  loading: state.loading,
  searchedBooks: state.searchedBooks,
  savedBooks: state.savedBooks,
})

export const mapDispatchToProps = (dispatch) => ({
  setError: (data) => dispatch(actions.setError(data)),
  setLoading: (data) => dispatch(actions.setLoading(data)),
  setBookSearch: (data) => dispatch(actions.setBookSearch(data)),
})

App.propTypes = {
  error: PropTypes.string,
  loading: PropTypes.bool,
  searchedBooks: PropTypes.array,
  savedBooks: PropTypes.array,
  setError: PropTypes.func,
  setLoading: PropTypes.func,
  setBookSearch: PropTypes.func,
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
