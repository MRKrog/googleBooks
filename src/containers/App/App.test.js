import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { setLoading, setError } from '../../actions/index';

describe('App', () => {

  describe('App Component', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(
        <App />
      )

      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockData)
      }))

    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

  });

  describe('mapStateToProps', () => {

  });

  describe('mapDispatchToProps', () => {

  });

});
