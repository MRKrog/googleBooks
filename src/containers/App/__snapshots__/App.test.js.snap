// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App Component should match the snapshot 1`] = `
<div
  className="App"
>
  <div
    className="marvel-device iphone-x"
  >
    <div
      className="notch"
    >
      <div
        className="camera"
      />
      <div
        className="speaker"
      />
    </div>
    <div
      className="top-bar"
    />
    <div
      className="sleep"
    />
    <div
      className="bottom-bar"
    />
    <div
      className="volume"
    />
    <div
      className="overflow"
    >
      <div
        className="shadow shadow--tr"
      />
      <div
        className="shadow shadow--tl"
      />
      <div
        className="shadow shadow--br"
      />
      <div
        className="shadow shadow--bl"
      />
    </div>
    <div
      className="inner-shadow"
    />
    <div
      className="screen"
    >
      <div
        className="Weather-Content"
      >
        <Route
          component={[Function]}
          path="/"
        />
        <Route
          path="/Search"
          render={[Function]}
        />
        <ConnectFunction />
        <ConnectFunction />
        <ConnectFunction />
        <NavLink
          className="Button-Container"
          to="/WeatherDetails"
        >
          <div
            className="Details-Button"
          >
            <i
              className="fas fa-plus-square"
            />
          </div>
        </NavLink>
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(WeatherDetails)",
              "type": [Function],
              Symbol(Symbol.hasInstance): [Function],
            }
          }
          exact={true}
          path="/WeatherDetails"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`App App Component should match the snapshot when loading is false 1`] = `
<div
  className="App"
>
  <div
    className="marvel-device iphone-x"
  >
    <div
      className="notch"
    >
      <div
        className="camera"
      />
      <div
        className="speaker"
      />
    </div>
    <div
      className="top-bar"
    />
    <div
      className="sleep"
    />
    <div
      className="bottom-bar"
    />
    <div
      className="volume"
    />
    <div
      className="overflow"
    >
      <div
        className="shadow shadow--tr"
      />
      <div
        className="shadow shadow--tl"
      />
      <div
        className="shadow shadow--br"
      />
      <div
        className="shadow shadow--bl"
      />
    </div>
    <div
      className="inner-shadow"
    />
    <div
      className="screen"
    >
      <div
        className="Weather-Content"
      >
        <Route
          component={[Function]}
          path="/"
        />
        <Route
          path="/Search"
          render={[Function]}
        />
        <ConnectFunction />
        <ConnectFunction />
        <ConnectFunction />
        <NavLink
          className="Button-Container"
          to="/WeatherDetails"
        >
          <div
            className="Details-Button"
          >
            <i
              className="fas fa-plus-square"
            />
          </div>
        </NavLink>
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(WeatherDetails)",
              "type": [Function],
              Symbol(Symbol.hasInstance): [Function],
            }
          }
          exact={true}
          path="/WeatherDetails"
        />
      </div>
    </div>
  </div>
</div>
`;
